Backends/CLX/clipboard.lisp

```diff
@@ -0,0 +1,227 @@
+(in-package :clim-clx)
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; CLX implementation of clipboard management
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+;;; Port mixin
+
+(defclass clx-clipboard-stored-object ()
+  ((content :initarg :content
+            :reader clipboard-stored-object-content)
+   (type    :initarg :type
+            :reader clipboard-stored-object-type)
+   (owner   :initarg :owner
+            :reader clipboard-stored-object-owner)))
+
+(defclass clx-clipboard-port-mixin ()
+  ((selection-content :initform nil
+                      :accessor selection-content)
+   (clipboard-content :initform nil
+                      :accessor clipboard-content)
+   (outstanding-request-pane :initform nil
+                             :accessor clipboard-outstanding-request-pane)
+   (outstanding-request-type  :initform nil
+                              :accessor clipboard-outstanding-request-type)
+   (outstanding-request-selection :initform nil
+                                  :accessor clipboard-outstanding-request-selection)))
+
+(defun find-stored-object (port selection)
+  (ecase selection
+    (:primary (selection-content port))
+    (:clipboard (clipboard-content port))))
+
+(defun set-stored-object (port selection value)
+  (ecase selection
+    (:primary (setf (selection-content port) value))
+    (:clipboard (setf (clipboard-content port) value))))
+
```

Comment:
`find-stored-object` and `set-stored-object` should have a shared lock (may be a slot in port, for instance clipboard-lock). they are accessed from both input processing thread and the application command loop(s). Sadly neither bordeaux-threads nor clim-sys defines read-write locks, because this would be a classic example.

```
(process-next-event-thread
 (:event :selection-notify
         (event-handler)
         (process-selection-notify)
         (process-string-reply)         ; not always strings!
         (sheet-event-queue clipboard-send-event))

 (:event :selection-request
         (event-handler)
         (process-selection-request)
         (find-stored-object))

 (:event :selection-clear
         (event-handler)
         (process-selection-clear)
         (set-stored-object))

 (:event :mixed-event-drag-with-key-modifiers
         (dispatch-event :around)
         (eos/shift-click)
         (pane-clear-markings)
         (clime:clear-selection)
         (clx-clear-clipboard)
         (find-stored-object))

 (:event :mixed-event-release
         (dispatch-event :around)
         (eos/shift-release)
         (clime:copy-to-selection)
         (clx-copy-to-clipboard)
         (set-stored-object)))

(command-loop-thread
 (:command (com-copy-text-to-clipboard) ; selection ->
           (clime:local-seleciton-content)
           (find-stored-object))
 (:command (com-copy-text-to-clipboard) ; -> clipboard
           (clime:copy-to-clipboard)
           (clx-copy-to-clipboard)
           (set-stored-object))
 (:command (com-copy-to-clipboard)
           (clime:copy-to-clipboard)
           (clx-copy-to-clipboard)
           (set-stored-object))
 (:drei (probably-handle-event)
        (push-start-and-copy)
        (clime:copy-to-clipboard)
        (clx-copy-to-clipboard)
        (set-stored-object)))
```

Reply:
this are simple accessors, so I don't think it is possible that incomplete state is read.